bam2depth.c

```diff
@@ -41,377 +43,883 @@ DEALINGS IN THE SOFTWARE.  */
+                kputc('\n', &dh->ks);
+                fputs(dh->ks.s, opt->out);
+            }
+            if (opt->all_pos && i < b->core.pos)
+                // Hole in middle of ref
+                zero_region(opt, dh, dh->ref, i, b->core.pos);
+
+            dh->ks.l = cur_l;
+            dh->last_output = b->core.pos;
         }
     }
-    if (optind == argc && !file_list)
-        return usage();
-
-    /* output file provided by user */
-    if (output_file != NULL && strcmp(output_file,"-")!=0) {
-        file_out = fopen( output_file, "w" );
-        if (file_out == NULL) {
-            print_error_errno("depth", "Cannot open \"%s\" for writing.", output_file);
-            return EXIT_FAILURE;
-        }
+
+    hts_pos_t end_pos = bam_endpos(b); // 0 based, 1 past end.
+    //printf("%d %d\n", (int)b->core.pos+1, (int)end_pos);
+
+    if (b->core.tid < dh->last_ref ||
+        (dh->last_ref == b->core.tid && end_pos < dh->last_output)) {
+        print_error_errno("depth", "Data is not position sorted");
+        return -1;
     }
 
+    // If needed, grow the circular buffer.
+    if (end_pos+1 - b->core.pos >= dh->size) {
+        size_t old_size = dh->size;
+        size_t old_hmask = hmask;
+        while (end_pos+1 - b->core.pos >= dh->size)
+            dh->size = dh->size ? 2*dh->size : 2048;
+        hmask = dh->size-1;
+        if (!dh->hist) {
+            dh->hist = calloc(dh->nfiles, sizeof(*dh->hist));
+            dh->end_pos = calloc(dh->nfiles, sizeof(*dh->end_pos));
+            if (!dh->hist || !dh->end_pos)
+                return -1;
+        }
+        for (n = 0; n < dh->nfiles; n++) {
+            int *hist = calloc(dh->size, sizeof(*dh->hist[n]));
+            if (!hist)
+                return -1;
 
-    // initialize the auxiliary data structures
-    if (file_list)
-    {
-        if (has_index_file) {
-            print_error("depth", "The -f option cannot be combined with -X");
-            return 1;
+            // Simple approach for now; copy over old histogram verbatim.
+            for (i = dh->last_output; i < dh->last_output + old_size; i++)
+                hist[i & hmask] = dh->hist[n][i & old_hmask];
+            free(dh->hist[n]);
+            dh->hist[n] = hist;
         }
-        if ( read_file_list(file_list,&nfiles,&fn) ) return EXIT_FAILURE;
-        n = nfiles;
-        argv = fn;
-        optind = 0;
     }
-    else if (has_index_file) { // Calculate # of input BAM files
-        if ((argc - optind) % 2 != 0) {
-            fprintf(stderr, "Error: Odd number of filenames detected! Each BAM file should have an index file\n");
-            return 1;
-        }
-        n = (argc - optind) / 2;
+
+    // Accumulate depth, based on CIGAR
+    uint32_t *cig = bam_get_cigar(b);
+    int ncig = b->core.n_cigar, j, k, spos = 0;
+
+    // Zero new (previously unseen) coordinates so increment works later.
+    // FIXME: do this for all files so output is easy,
```

Comment:
Since you have decided on which case to use, do you want to remove the FIXME?

Reply:
Good idea.